src/psbt.c

```diff
@@ -1,44 +1,72 @@
 #include "internal.h"
 
 #include "ccan/ccan/base64/base64.h"
-#include "ccan/ccan/build_assert/build_assert.h"
 
-#include <include/wally_crypto.h>
+#include <include/wally_elements.h>
 #include <include/wally_script.h>
-#include <include/wally_transaction.h>
 #include <include/wally_psbt.h>
 
 #include <limits.h>
 #include <stdbool.h>
 #include "transaction_shared.h"
+#include "psbt_int.h"
 #include "script_int.h"
 #include "script.h"
+#include "pullpush.h"
 
-const uint8_t WALLY_PSBT_MAGIC[5] = {'p', 's', 'b', 't', 0xff};
+static const uint8_t WALLY_PSBT_MAGIC[5] = {'p', 's', 'b', 't', 0xff};
+static const uint8_t WALLY_ELEMENTS_PSBT_MAGIC[5] = {'p', 's', 'e', 't', 0xff};
 
+#ifdef BUILD_ELEMENTS
+static const uint8_t WALLY_ELEMENTS_ID[8] = {'e', 'l', 'e', 'm', 'e', 'n', 't', 's'};
+static const size_t WALLY_ELEMENTS_ID_LEN = 8;
+#endif /* BUILD_ELEMENTS */
 
-static bool pubkey_is_compressed(unsigned char pubkey[EC_PUBLIC_KEY_UNCOMPRESSED_LEN]) {
-    return pubkey[0] == 0x02 || pubkey[0] == 0x03;
+static bool pubkey_is_compressed(const unsigned char pub_key[EC_PUBLIC_KEY_UNCOMPRESSED_LEN]) {
+    return pub_key[0] == 0x02 || pub_key[0] == 0x03;
 }
 
+static size_t get_pubkey_len(const unsigned char *pub_key, size_t pub_key_len)
+{
+    if (!pub_key_len)
+        return 0;
+    return pubkey_is_compressed(pub_key) ? EC_PUBLIC_KEY_LEN : EC_PUBLIC_KEY_UNCOMPRESSED_LEN;
+}
+
+static bool is_valid_pubkey_len(size_t len)
+{
+    return len == EC_PUBLIC_KEY_UNCOMPRESSED_LEN || len == EC_PUBLIC_KEY_LEN;
+}
+
+static bool is_valid_optional_pubkey_len(size_t len)
+{
+    return !len || is_valid_pubkey_len(len);
+}
 
-int wally_keypath_map_init_alloc(size_t alloc_len, struct wally_keypath_map **output)
+static bool is_equal_pubkey(const unsigned char *l, size_t l_len, const unsigned char *r, size_t r_len)
+{
+    if (!l || !r)
+        return l == r; /* Equal if both NULL only */
+
+    return l_len == r_len && memcmp(l, r, l_len) == 0;
+}
+
+int wally_keypath_map_init_alloc(size_t allocation_len, struct wally_keypath_map **output)
```

Comment:
For keypaths, partial sigs and unknowns maps we have some functions (init alloc, free, find, add, replace) which are almost identical. Do you think it's worth to abstract them somehow and share some code?

Reply:
Yes, absolutely!

The reality is that the initially implemented PSBT interface isn't fit for purpose, and will change more beyond the changes here. As you point out for example, partial sigs and unknowns maps are exactly the same (key->value maps) and keypath could be implemented using the same code if the keypath is stored as it is in the PSBT file (key->value with the fingerprint as the first element in the path). That would allow us to share the code for all 3 maps and get rid of 2 of the types and associated wrapping code (and fix the naming inconsistencies and plural confusion in the type/function names).

Also, PSBT is the 2nd library addition that requires structure manipulation and so some of the code that supports this in transaction.c should be generalised so it can be reused when we add the 3rd (descriptors or whatever).

However - These changes will break the PSBT interface completely. My view is that no one can have been using it previously due to the bugs, as evidenced by c-lightning having to rewrite and fix large parts of the code in order to get things working (and they still have a bunch of workarounds in their tree after moving to this branch).

So, once this PR is merged I want to push the bigger interface changes before the next release and just document the change in CHANGES.md as 'the PSBT interface has changed and can now be considered final'. That deserves its own PR though so any bike-shedding over the interface can happen there.